File: php/src/Hqc/GF2.php
<?php

declare(strict_types=1);

namespace PQC\Hqc;

/**
 * GF(2) polynomial arithmetic using GMP for arbitrary-precision bit operations.
 *
 * Polynomials are represented as GMP integers where bit i is the coefficient of x^i.
 * Arithmetic is in GF(2)[x] / (x^n - 1).
 */
final class GF2
{
    /**
     * Add two GF(2) polynomials (XOR).
     */
    public static function add(\GMP $a, \GMP $b): \GMP
    {
        return gmp_xor($a, $b);
    }

    /**
     * Set bit at position $pos.
     */
    public static function setBit(\GMP $v, int $pos): \GMP
    {
        return gmp_or($v, gmp_pow(gmp_init(2), $pos));
    }

    /**
     * Get bit at position $pos.
     */
    public static function getBit(\GMP $v, int $pos): int
    {
        return gmp_testbit($v, $pos) ? 1 : 0;
    }

    /**
     * Hamming weight (number of set bits).
     */
    public static function weight(\GMP $v): int
    {
        return gmp_popcount($v);
    }

    /**
     * Mask to n bits: clear all bits >= n.
     */
    public static function mask(\GMP $v, int $n): \GMP
    {
        $m = gmp_sub(gmp_pow(gmp_init(2), $n), gmp_init(1));
        return gmp_and($v, $m);
    }

    /**
     * Multiply two GF(2) polynomials modulo (x^n - 1).
     *
     * Uses schoolbook multiplication with reduction.
     */
    public static function mulMod(\GMP $a, \GMP $b, int $n): \GMP
    {
        // Full carryless multiplication
        $result = gmp_init(0);
        $bb = $b;
        $shift = 0;

        while (gmp_cmp($bb, gmp_init(0)) > 0) {
            if (gmp_testbit($bb, 0)) {
                $result = gmp_xor($result, gmp_mul(gmp_init(1), self::shiftLeft($a, $shift)));
            }
            $bb = self::shiftRight($bb, 1);
            $shift++;
        }

        // Reduce modulo x^n - 1
        $result = self::reduceMod($result, $n);
        return $result;
    }

    /**
     * Reduce polynomial modulo (x^n - 1).
     * Since x^n = 1, we XOR bits above position n back at position 0.
     */
    public static function reduceMod(\GMP $a, int $n): \GMP
    {
        $maskVal = gmp_sub(gmp_pow(gmp_init(2), $n), gmp_init(1));
        $result = gmp_and($a, $maskVal);
        $a = self::shiftRight($a, $n);

        while (gmp_cmp($a, gmp_init(0)) > 0) {
            $result = gmp_xor($result, gmp_and($a, $maskVal));
            $a = self::shiftRight($a, $n);
        }

        return $result;
    }

    /**
     * Convert GMP polynomial to bytes (little-endian).
     */
    public static function toBytes(\GMP $v, int $nBytes): string
    {
        $result = str_repeat("\x00", $nBytes);
        $bytes = [];
        for ($i = 0; $i < $nBytes; $i++) {
            $byte = 0;
            for ($b = 0; $b < 8; $b++) {
                if (gmp_testbit($v, $i * 8 + $b)) {
                    $byte |= (1 << $b);
                }
            }
            $bytes[] = $byte;
        }
        return pack('C*', ...$bytes);
    }

    /**
     * Convert bytes to GMP polynomial (little-endian).
     */
    public static function fromBytes(string $data, int $nBits): \GMP
    {
        $result = gmp_init(0);
        $len = strlen($data);
        for ($i = $len - 1; $i >= 0; $i--) {
            $result = gmp_or(
                self::shiftLeft($result, 8),
                gmp_init(ord($data[$i]))
            );
        }
        return self::mask($result, $nBits);
    }

    /**
     * Constant-time equality check. Returns 1 if equal, 0 otherwise.
     */
    public static function equal(\GMP $a, \GMP $b): int
    {
        return (gmp_cmp(gmp_xor($a, $b), gmp_init(0)) === 0) ? 1 : 0;
    }

    /**
     * Shift left by $n bits (GF(2) polynomial << n).
     */
    public static function shiftLeft(\GMP $v, int $n): \GMP
    {
        if ($n === 0) {
            return $v;
        }
        // GMP doesn't have native shift, use multiplication by 2^n
        return gmp_mul($v, gmp_pow(gmp_init(2), $n));
    }

    /**
     * Shift right by $n bits (GF(2) polynomial >> n).
     */
    public static function shiftRight(\GMP $v, int $n): \GMP
    {
        if ($n === 0) {
            return $v;
        }
        return gmp_div_q($v, gmp_pow(gmp_init(2), $n));
    }

    /**
     * Create GMP zero.
     */
    public static function zero(): \GMP
    {
        return gmp_init(0);
    }
}

File: python/hqc/gf2.py
"""GF(2) polynomial arithmetic using Python arbitrary-precision integers.

Each polynomial is represented as a Python int where bit i is the coefficient
of x^i. This gives us XOR for addition, bit manipulation for coefficients,
and arbitrary precision for free.
"""


def gf2_add(a: int, b: int) -> int:
    """Add two GF(2) polynomials (XOR)."""
    return a ^ b


def gf2_get_bit(a: int, i: int) -> int:
    """Get the coefficient of x^i in polynomial a."""
    return (a >> i) & 1


def gf2_set_bit(a: int, i: int) -> int:
    """Return a with the coefficient of x^i set to 1."""
    return a | (1 << i)


def gf2_clear_bit(a: int, i: int) -> int:
    """Return a with the coefficient of x^i set to 0."""
    return a & ~(1 << i)


def gf2_weight(a: int) -> int:
    """Compute the Hamming weight (number of 1 bits) of a GF(2) polynomial."""
    return bin(a).count('1')


def gf2_mul_mod(a: int, b: int, n: int) -> int:
    """Multiply two GF(2) polynomials modulo (x^n - 1).

    This computes a * b in GF(2)[x] / (x^n - 1).
    Uses schoolbook multiplication with reduction.
    """
    result = 0
    # Iterate over set bits of b
    bb = b
    shift = 0
    while bb:
        if bb & 1:
            result ^= (a << shift)
        bb >>= 1
        shift += 1

    # Reduce modulo x^n - 1 (i.e., x^n = 1, so x^(n+k) = x^k)
    result = _reduce_mod_xn_minus_1(result, n)
    return result


def _reduce_mod_xn_minus_1(a: int, n: int) -> int:
    """Reduce polynomial a modulo (x^n - 1).

    Since x^n = 1 in the quotient ring, any term x^(n+k) wraps to x^k.
    """
    mask = (1 << n) - 1
    result = a & mask
    a >>= n
    while a:
        result ^= (a & mask)
        a >>= n
    return result


def gf2_sparse_mul_mod(a: int, positions: list, n: int) -> int:
    """Multiply GF(2) polynomial a by a sparse polynomial given by its support positions.

    The sparse polynomial b = sum(x^pos for pos in positions).
    Result is computed mod (x^n - 1).
    """
    result = 0
    for pos in positions:
        result ^= _rotate_left(a, pos, n)
    return result


def _rotate_left(a: int, shift: int, n: int) -> int:
    """Rotate polynomial a left by shift positions modulo x^n - 1."""
    if shift == 0:
        return a
    shift = shift % n
    mask = (1 << n) - 1
    return ((a << shift) | (a >> (n - shift))) & mask


def gf2_to_bytes(a: int, n: int) -> bytes:
    """Serialize a GF(2) polynomial of degree < n to bytes (little-endian bit order).

    The number of bytes is ceil(n / 8).
    """
    num_bytes = (n + 7) // 8
    result = bytearray(num_bytes)
    for i in range(num_bytes):
        result[i] = (a >> (i * 8)) & 0xFF
    return bytes(result)


def gf2_from_bytes(data: bytes, n: int) -> int:
    """Deserialize a GF(2) polynomial from bytes (little-endian bit order).

    Only the first n bits are used.
    """
    result = int.from_bytes(data, byteorder='little')
    mask = (1 << n) - 1
    return result & mask


def gf2_random_fixed_weight(rand_bytes: bytes, n: int, w: int) -> int:
    """Generate a random GF(2) polynomial of length n with exactly w nonzero coefficients.

    Uses Fisher-Yates-like rejection sampling from random bytes.
    The rand_bytes should contain enough randomness (at least 4*w bytes recommended).
    """
    # Parse random bytes into 32-bit values for position sampling
    positions = set()
    offset = 0
    while len(positions) < w:
        if offset + 4 > len(rand_bytes):
            raise ValueError("Not enough random bytes for weight generation")
        val = int.from_bytes(rand_bytes[offset:offset+4], byteorder='little')
        offset += 4
        pos = val % n
        positions.add(pos)

    result = 0
    for pos in positions:
        result = gf2_set_bit(result, pos)
    return result


def gf2_random_fixed_weight_from_shake(shake, n: int, w: int) -> int:
    """Generate a random GF(2) polynomial of length n with exactly w nonzero coefficients.

    Uses a SHAKE-256 instance for randomness via rejection sampling.
    """
    positions = []
    seen = set()
    # We need to sample w distinct positions from [0, n)
    # Use 4-byte chunks from SHAKE for each candidate
    while len(positions) < w:
        rand_bytes = shake.digest(4 + len(positions) * 0)
        # Actually we need to squeeze incrementally. Use a counter approach.
        break

    # Better approach: squeeze enough bytes at once and parse
    # For rejection sampling, we need roughly w * 4 bytes, but may need more
    # Squeeze a generous amount
    needed = w * 8  # generous estimate
    data = shake.digest(needed)

    positions = []
    seen = set()
    offset = 0
    while len(positions) < w:
        if offset + 4 > len(data):
            # Need more randomness - squeeze more
            needed *= 2
            data = shake.digest(needed)
        val = int.from_bytes(data[offset:offset+4], byteorder='little')
        offset += 4
        pos = val % n
        if pos not in seen:
            seen.add(pos)
            positions.append(pos)

    result = 0
    for pos in positions:
        result = gf2_set_bit(result, pos)
    return result
